[Khang Hou] iP

tankh99 opened the pull request:
### Baron
Behold, Baron and his new features, including
1. Malicious sarcasm
2. Database storage feaatures

Additionally he can
- Remember your face and name
- Know where you live

Just watch out for this line while reading the codebase

```java
System.out.println("If you see this, you die");
```

Future enhancements include
- [ ] OpenAI integration
- [ ] Actually being able to control you like the peasant you are as it is his royal right to do so :smile:

> I want to let you know, that you're being disrespectful to a noble. Guards, send him for execution

Is just one of the quotes that Baron will NOT say to you because I haven't implemented it!

[Visit this page to get a virus](https://example.com/)

But if you ever get a virus, just write this line in your terminal `stop virus`

I also ~~don't~~ _have_ a **virus**

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

tankh99 committed:
Level-0

tankh99 committed:
Level-1

tankh99 committed:
Level-2

tankh99 committed:
Level-3

tankh99 committed:
Level-4

tankh99 committed:
fix: Level-4 print formatting

tankh99 committed:
A-TextUITesting

tankh99 committed:
fix: A-TextUITesting add stop to program test

tankh99 committed:
refactor: move models into models folder

tankh99 committed:
feat: add input handlers for each model

tankh99 committed:
refactor: move input handling to separate file

tankh99 committed:
refactor: fix value parsing function

tankh99 committed:
refactor: fix string formatting

tankh99 committed:
refactor: change exception

tankh99 committed:
Level-5

tankh99 committed:
feat: add delete
refactor: use const for commands
refactor: make const task names public

tankh99 committed:
chore: update documentation and future goals

tankh99 committed:
Level-6

tankh99 committed:
A-Enums

tankh99 committed:
refactor: move Daos to Dao package

tankh99 committed:
feat: add create and read for database

tankh99 committed:
fix: fix tests

tankh99 committed:
feat: add tasktype enum

tankh99 committed:
feat: add fileutils

tankh99 committed:
refactor: use task type enum for DAO names
refactor: use path as parameter instead of string for db operations
feat: add update by id

tankh99 committed:
refactor: add TaskDao to abstract mark functionality
refactor: use task type classification

tankh99 committed:
refactor: return id when creating

previously, creating 2 tasks in the same session didn't actually increment their ID, so marking would be incorrect. we now return id when creating, and we set the id so that id is up to date

tankh99 committed:
refactor: add function to TodoDao

Abstracted add duplicate logic into TodoDao

tankh99 committed:
fix: index issue when updating

tankh99 committed:
chore: ignore db files in git

tankh99 committed:
refactor: move UI code to UIManager

tankh99 committed:
refactor: move db-specific logic out of fileutils into database

tankh99 committed:
Level-7

fix: delete logic

tankh99 committed:
Level-8

feat: add DateUtils to parse date only for deadlines

tankh99 committed:
Merge pull request #1 from tankh99/Level-7

Level 7

tankh99 committed:
Merge pull request #2 from tankh99/Level-8

Level-8

tankh99 committed:
Merge pull request #3 from tankh99/add-gradle-support

Add gradle support

tankh99 committed:
chore: add tests for utils files

tankh99 committed:
chore: add javadoc comments
refactor: splitDataString not trimming output and remove redundant method

tankh99 committed:
refactor: nest entire project into rªoot folder

tankh99 committed:
refactor: update runtest.sh

tankh99 committed:
chore: add jar

tankh99 committed:
Merge pull request #4 from tankh99/A-Packages

A packages

tankh99 committed:
Merge pull request #5 from tankh99/A-Jar

A-Jar

tankh99 committed:
A-Javadoc

chore: add havadocs to most public functions and class headers

tankh99 committed:
A-CodingStandard

- reformatted code base according to modified intellij coding standard

tankh99 committed:
Level-9

tankh99 committed:
Merge pull request #6 from tankh99/A-JavaDoc

A-JavaDoc

tankh99 committed:
chore: add javadoc for Level-9

tankh99 committed:
Merge pull request #7 from tankh99/Level-9

Level-9

tankh99 committed:
fix: rebase errors

tankh99 committed:
chore: reformat code

tankh99 committed:
Merge pull request #8 from tankh99/A-CodingStandard

A-CodingStandard

tankh99 committed:
fix: missing functions from rebase

jaejayrome commented on src/main/java/baron/Utils/DateUtils.java:
```diff
@@ -0,0 +1,38 @@
+ */
+public class DateUtils {
+
+    // 03/21/1999 1600
```
I like how you have included an example DateTime format to allow other developers to have an understanding of the format required!

jaejayrome commented on src/main/java/baron/Baron.java:
```diff
@@ -0,0 +1,23 @@
+import java.util.Scanner;
+
+public class Baron {
+    public static void main(String[] args) {
```
I feel that there could be a Javadoc comment to mention how this method would be the main entrypoint of your application?

jaejayrome commented on src/main/java/baron/Models/Deadline.java:
```diff
@@ -0,0 +1,46 @@
+
+    private final LocalDateTime deadline;
+
+    public Deadline(String name, LocalDateTime deadline) {
+        super(name);
+        this.deadline = deadline;
+    }
+
+    public Deadline(int id, String name, boolean isDone, LocalDateTime deadline) {
+        super(id, name, isDone);
+        this.deadline = deadline;
+    }
```
I personally like how you implemented method overloading but I feel that you could include Javadoc comments for constructors as well :)

jaejayrome commented on src/main/java/baron/Models/Event.java:
```diff
@@ -0,0 +1,54 @@
+        return event;
+    }
+
+    public String getStartDate() {
+        return this.startDate;
+    }
+
+    public void setStartDate(String startDate) {
+        this.startDate = startDate;
+    }
```
Very good adherence to the coding style thus far! It would be better if you could include Javadoc comments for Getters and Setters as well especially for novice programmers that may be new to Java!

tankh99 replied:
I think getters and setters are generally accepted to not need JavaDocs? Correct me if I'm wrong

jaejayrome commented on src/main/java/baron/Utils/DateUtils.java:
```diff
@@ -0,0 +1,38 @@
+        try {
+            return LocalDateTime.parse(dateString, INPUT_FORMATTER);
+        } catch (DateTimeParseException ex) {
+            throw new IllegalArgumentException("Invalid date string given. Expected format is d/M/yyyy HHmm (e.g. 21/10/1999 1600)");
+        }
```
I like how you incorporated the IllegalArgumentException class which would make it more apparent to users what the exact exception would be given your custom error message!

jaejayrome commented on src/test/java/baron/Utils/DateUtilsTest.java:
```diff
@@ -0,0 +1,62 @@
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+public class DateUtilsTest {
+    @Test
+    public void parseDateString_correctDateFormat_dateReturned() {
+        String dateString = "9/10/1999 1600";
+        LocalDateTime actual = DateUtils.parseDate(dateString);
+        LocalDateTime expected = LocalDateTime.of(1999, 10, 9, 16, 0, 0);
+        assertEquals(actual, expected);
+    }
+
+    @Test
+    public void parseDateString_incorrectDateFormat_dateReturned() {
+        // Invalid day
+        assertThrows(IllegalArgumentException.class, () -> {
+            String dateString = "32/10/1999 1600";
+            LocalDateTime actual = DateUtils.parseDate(dateString);
+        });
+
+        // Invalid month
+        assertThrows(IllegalArgumentException.class, () -> {
+            String dateString = "1/13/1999 1600";
+            LocalDateTime actual = DateUtils.parseDate(dateString);
+        });
+
+        // Invalid year
+        assertThrows(IllegalArgumentException.class, () -> {
+            String dateString = "1/13/199 1600";
+            LocalDateTime actual = DateUtils.parseDate(dateString);
+        });
+    }
```
I like how your test cases are clear and elaborate and most importantly cover all edge cases that might happen, the detailing is elaborate and concise! 👍

jaejayrome commented on src/test/java/baron/Utils/FileUtilsTest.java:
```diff
@@ -0,0 +1,26 @@
+public class FileUtilsTest {
+    public static final String BASE_DIR = System.getProperty("user.dir") + "/src/test/java/baron.Database";
+
+    public FileUtilsTest() {
+
+    }
```
I like how you have added an empty constructor since it may not be apparent to people who are new to Java that every class would be coupled with a default empty constructor 💯

jaejayrome reviewed (commented):
LGTM! Just a few nits have been spotted but mostly looks fine 🥇

tankh99 committed:
feat: Add JavaFX skeleton code

This commit adds DialogBox, MainWindow and basic GUI, including images, to imitate a chatbox

tankh99 committed:
Level-10

feat: Bot now uses responses from the main chatbot logic
refactor: UIManager methods now return string instead of printing out directly

tankh99 committed:
A-CheckStyle

chore: setup checkstyle

tankh99 committed:
Merge pull request #10 from tankh99/A-CheckStyle

A-CheckStyle

tankh99 committed:
Merge pull request #9 from tankh99/Level-10

Level 10

tankh99 committed:
refactor: rename directories to lowercase

tankh99 committed:
chore: Fix all checkstyle transgressions

tankh99 committed:
Merge pull request #11 from tankh99/FixLint

Fix linting

evedaktyl commented on src/main/java/baron/database/Database.java:
```diff
@@ -0,0 +1,148 @@
+     * @param line     Line to append to the end of file
+     * @return the id of the newly appended line, which is actually just lineNumber - 1
+     */
+    public static long create(Path filePath, String line) {
```
Perhaps we could use a more appropriate name for this method, such as write/writeToFile()?

evedaktyl commented on src/main/java/baron/utils/FileUtils.java:
```diff
@@ -0,0 +1,44 @@
+/**
+ * In charge of managing simple file IO such as reading and writing
+ */
+public class FileUtils {
```
I like that you have this extraction of file IO operations.

evedaktyl commented on src/main/java/baron/managers/TaskManager.java:
```diff
@@ -0,0 +1,168 @@
+    /**
+     * Prints out a list of tasks that match the search term (fuzzy search)
+     *
+     * @param input user input
```
Perhaps we could be consistent in starting the param descriptors with a capital letter to enhance readability.

evedaktyl commented on src/main/java/baron/managers/UiManager.java:
```diff
@@ -0,0 +1,81 @@
+     * @param task  Task added
+     * @param count count of tasks currently in list after modification
+     */
+    public static String delete(Task task, int count) {
```
Could we consider naming this method as showDeleteMessage() to enhance clarity?

evedaktyl reviewed (commented):
Overall code is readable and neat. Apart from a few method names that could benefit from slightly more clarity, code quality is great!

tankh99 committed:
chore: Create jar file

tankh99 committed:
feat: Add simple assertions

tankh99 committed:
Fix find not being case-insensitive

Find command also now properly handles empty results

tankh99 committed:
refactor: Rename enums

1.Rename Commands enum file to be singular in accordance to naming convention.
2. Renamed variables in TaskType enum file
3. Add Argument enum to represent different argument types for user input

tankh99 committed:
Add punctuation to end of javadocs and clean up code

- removed unused methods and imports
- moved getCommand to StringUtils

tankh99 committed:
Merge pull request #12 from tankh99/A-Assertions

feat: Add simple assertions

tankh99 committed:
Merge pull request #13 from tankh99/A-CodeQuality

A-CodeQuality

tankh99 committed:
Create Model abstraction and Client model

- Created the Client model.
- Refactored Task model to inherit from Model
- Refactored long id to be int id to be consistent with other methods

tankh99 committed:
Create BaseDao

- Refactor Model to BaseModel
- Created BaseDao abstraction which will help ensure all Dao classes share similar logic

tankh99 committed:
Fix argument typo

tankh99 committed:
Allow create client

- refactor filePath param to name
- add Client getters and setters
- create Client

tankh99 committed:
Fix issues nad create separator CONST

tankh99 committed:
Add command to list all stored clients

tankh99 committed:
Allow deleting clients

tankh99 committed:
Merge pull request #14 from tankh99/D-Clients

D clients

tankh99 committed:
refactor: Rename list command to tasks

- Also refactored command enum

tankh99 committed:
Update README and title of application

tankh99 committed:
Merge pull request #15 from tankh99/A-UserGuide

A-UserGuide

tankh99 committed:
Add Ui image

tankh99 committed:
Reactor README from root to docs folder

tankh99 committed:
Add manifest for jar and update README

tankh99 committed:
Fix gradle build issues

tankh99 committed:
Update Javadoc comments

tankh99 committed:
Fix bad build.gradle configs and refactored directory

tankh99 committed:
Add help command

tankh99 committed:
Rename send button

tankh99 committed:
Fixed UI problems

tankh99 committed:
Fix: negative indexing in string utils

tankh99 committed:
refactor: use single function to get item index

tankh99 committed:
chore: use single-class imports

tankh99 committed:
refactor: use ListView instead of ScrollPane for performance

tankh99 committed:
fix: stretch height of list view to fit
